### ISSUE TITLE
[sbt0.11.2] dependsOn(x) libraries not included in sbt-idea export where 'x' is multi-module project

### ISSUE THREAD

**Issue by opyate** [mpeltonen/sbt-idea#126]
Hi Mikko,

As discussed in issue 62, here is an early-stage project of mine which can be used as a reproducible test case:

https://github.com/opyate/grafen (the `sbtidea` branch!)
It depends on https://github.com/opyate/scala-graph (scala-graph is a multi-module project)

Here is a first run, and you can see the dependent cloning into a staging environment, but it does not get created as a library in the IDEA project:

``` bash
$ ./sbt
[info] Loading global plugins from /Users/juanuys/.sbt/plugins/project
[info] Loading global plugins from /Users/juanuys/.sbt/plugins
[info] Loading project definition from /Users/juanuys/Documents/src/github/graph-projects/grafen/project
[info] Compiling 1 Scala source to /Users/juanuys/Documents/src/github/graph-projects/grafen/project/target/scala-2.9.1/sbt-0.11.2/classes...
Cloning into '/Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb'...
[info] Loading project definition from /Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb/project
[info] Updating {file:/Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb/project/}default-30ac3f...
[info] Resolving default#default-8854bf;0.0 ...
[info] Resolving com.github.mpeltonen#sbt-idea;1.0.0 ...
[info] Resolving commons-io#commons-io;2.0.1 ...
[info] Resolving com.typesafe.sbteclipse#sbteclipse-plugin;2.0.0-M3 ...
[info] Resolving com.typesafe.sbteclipse#sbteclipse-core;2.0.0-M3 ...
[info] Resolving org.scalaz#scalaz-core_2.9.1;6.0.3 ...
[info] Resolving org.scala-lang#scala-library;2.9.1 ...
[info] Resolving com.eed3si9n#sbt-appengine;0.3.1 ...
[info] Resolving com.github.siasia#xsbt-web-plugin_2.9.1;0.11.2-0.2.10 ...
[info] Resolving com.github.siasia#plugin-commons_2.9.1;0.11.2-0.1 ...
[info] Resolving me.lessis#ls-sbt;0.1.0 ...
[info] Resolving me.lessis#ls_2.9.1;0.1.0 ...
[info] Resolving com.codahale#jerkson_2.9.1;0.5.0 ...
[info] Resolving org.codehaus.jackson#jackson-core-asl;[1.9.0,2.0.0) ...
[info] Resolving org.codehaus.jackson#jackson-mapper-asl;[1.9.0,2.0.0) ...
[info] Resolving net.databinder#dispatch-http_2.9.1;0.8.6 ...
[info] Resolving net.databinder#dispatch-core_2.9.1;0.8.6 ...
[info] Resolving org.apache.httpcomponents#httpclient;4.1.2 ...
[info] Resolving org.apache.httpcomponents#httpcore;4.1.2 ...
[info] Resolving commons-logging#commons-logging;1.1.1 ...
[info] Resolving commons-codec#commons-codec;1.4 ...
[info] Resolving net.databinder#dispatch-futures_2.9.1;0.8.6 ...
[info] Resolving org.scala-tools.sbt#sbt_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#main_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#actions_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#classfile_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#io_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#control_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#interface;0.11.2 ...
[info] Resolving org.scala-tools.sbt#logging_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#process_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#classpath_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#launcher-interface_2.9.1;0.11.2 ...
[info] Resolving org.scala-lang#scala-compiler;2.9.1 ...
[info] Resolving org.scala-tools.sbt#incremental-compiler_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#collections_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#api_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#persist_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbinary#sbinary_2.9.0;0.4.0 ...
[info] Resolving org.scala-tools.sbt#compile_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#ivy_2.9.1;0.11.2 ...
[info] Resolving org.apache.ivy#ivy;2.2.0 ...
[info] Resolving com.jcraft#jsch;0.1.31 ...
[info] Resolving commons-httpclient#commons-httpclient;3.1 ...
[info] Resolving commons-logging#commons-logging;1.0.4 ...
[info] Resolving commons-codec#commons-codec;1.2 ...
[info] Resolving org.scala-tools.sbt#completion_2.9.1;0.11.2 ...
[info] Resolving jline#jline;0.9.94 ...
[info] Resolving org.scala-tools.sbt#run_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#task-system_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#tasks_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#tracking_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#cache_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#testing_2.9.1;0.11.2 ...
[info] Resolving org.scala-tools.testing#test-interface;0.5 ...
[info] Resolving org.scala-tools.sbt#compiler-interface;0.11.2 ...
[info] Resolving org.scala-tools.sbt#precompiled-2_8_1;0.11.2 ...
[info] Resolving org.scala-tools.sbt#precompiled-2_8_0;0.11.2 ...
[info] Resolving org.scala-tools.sbt#precompiled-2_9_0;0.11.2 ...
[info] Done updating.
[info] Compiling 1 Scala source to /Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb/project/target/scala-2.9.1/sbt-0.11.2/classes...
[info] Set current project to grafen (in build file:/Users/juanuys/Documents/src/github/graph-projects/grafen/)
[info] Updating {git://github.com/opyate/scala-graph.git}Graph-all...
[info] Resolving org.scala-lang#scala-library;2.9.1 ...
[info] Resolving junit#junit;4.8.2 ...
[info] Resolving org.scalatest#scalatest_2.9.0;1.6.1 ...
[info] Done updating.
[info] Updating {file:/Users/juanuys/Documents/src/github/graph-projects/grafen/}grafen...
[info] Resolving scalax.collection.graph#graph-for-scala_2.9.1;1.3.2 ...
[info] Resolving org.scala-lang#scala-library;2.9.1 ...
[info] Resolving net.databinder#unfiltered-filter_2.9.1;0.5.3 ...
[info] Resolving net.databinder#unfiltered_2.9.1;0.5.3 ...
[info] Resolving net.databinder#unfiltered-util_2.9.1;0.5.3 ...
[info] Resolving commons-codec#commons-codec;1.4 ...
[info] Resolving net.databinder#unfiltered-spec_2.9.1;0.5.3 ...
[info] Resolving net.databinder#unfiltered-jetty_2.9.1;0.5.3 ...
[info] Resolving org.eclipse.jetty#jetty-webapp;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-xml;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-util;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-servlet;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-security;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-server;7.5.4.v20111024 ...
[info] Resolving javax.servlet#servlet-api;2.5 ...
[info] Resolving org.eclipse.jetty#jetty-continuation;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-http;7.5.4.v20111024 ...
[info] Resolving org.eclipse.jetty#jetty-io;7.5.4.v20111024 ...
[info] Resolving net.databinder#unfiltered-netty-server_2.9.1;0.5.3 ...
[info] Resolving net.databinder#unfiltered-netty_2.9.1;0.5.3 ...
[info] Resolving org.jboss.netty#netty;3.2.5.Final ...
[info] Resolving org.scala-tools.testing#specs_2.9.1;1.6.9 ...
[info] Resolving org.scala-lang#scala-compiler;2.9.1 ...
[info] Resolving net.databinder#dispatch-mime_2.9.1;0.8.6 ...
[info] Resolving net.databinder#dispatch-core_2.9.1;0.8.6 ...
[info] Resolving org.apache.httpcomponents#httpclient;4.1.2 ...
[info] Resolving org.apache.httpcomponents#httpcore;4.1.2 ...
[info] Resolving commons-logging#commons-logging;1.1.1 ...
[info] Resolving org.apache.httpcomponents#httpmime;4.1.2 ...
[info] Resolving org.apache.james#apache-mime4j;0.7.1 ...
[info] Resolving org.apache.james#apache-mime4j-core;0.7.1 ...
[info] Resolving org.apache.james#apache-mime4j-examples;0.7.1 ...
[info] Resolving org.apache.james#apache-mime4j-dom;0.7.1 ...
[info] Resolving org.apache.james#apache-mime4j-storage;0.7.1 ...
[info] Resolving org.apache.james#apache-mime4j-benchmark;0.7.1 ...
[info] Resolving commons-io#commons-io;1.4 ...
[info] Resolving net.databinder#dispatch-http_2.9.1;0.8.6 ...
[info] Resolving net.databinder#dispatch-futures_2.9.1;0.8.6 ...
[info] Resolving junit#junit;4.8.2 ...
[info] Resolving org.scalatest#scalatest_2.9.0;1.6.1 ...
[info] Resolving javax.servlet#servlet-api;2.3 ...
[info] Resolving org.eclipse.jetty#jetty-webapp;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-xml;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-util;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-servlet;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-security;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-server;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-continuation;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-http;7.4.5.v20110725 ...
[info] Resolving org.eclipse.jetty#jetty-io;7.4.5.v20110725 ...
[info] Done updating.

```

I generate the IDEA project:

``` bash
> gen-idea no-classifiers
[info] Trying to create an Idea module grafen
[info] Excluding folder target
[info] Created /Users/juanuys/Documents/src/github/graph-projects/grafen/.idea/IdeaProject.iml
[info] Created /Users/juanuys/Documents/src/github/graph-projects/grafen/.idea
[info] Excluding folder /Users/juanuys/Documents/src/github/graph-projects/grafen/target
[info] Created /Users/juanuys/Documents/src/github/graph-projects/grafen/.idea_modules/grafen.iml
[info] Created /Users/juanuys/Documents/src/github/graph-projects/grafen/.idea_modules/project.iml

```

Also, when opening the project, and trying `⌘N` to open, say, `Constrained`, it does not show one for `scalax.collection`. I.e. the scala-graph project is clearly not being created.

Thanks,
Juan

**Comment by mpeltonen** [comment]
You need to add also 'aggregate(scalaGraph)' to get gen-idea run for it too. Anyway, it seems to run only on root project, and only Graph-all.iml -module gets created.

**Comment by opyate** [comment]
Reproducible case now generates the following `projectList`, which correctly includes the root `grafen`, the dependency `Graph-all`, which in turn has `Graph-core` and `Graph-json` modules:

```
List(
  (ProjectRef(file:/Users/juanuys/Documents/src/github/graph-projects/grafen/,grafen),
  Project(
    id: grafen, 
    base: /Users/juanuys/Documents/src/github/graph-projects/grafen, 
    aggregate: List(ProjectRef(git://github.com/opyate/scala-graph.git,Graph-all)), 
    dependencies: List(
      ResolvedClasspathDependency(ProjectRef(git://github.com/opyate/scala-graph.git,Graph-all),None)), 
    delegates: List(), 
    configurations: List(compile, runtime, test, provided, optional))),

  (ProjectRef(git://github.com/opyate/scala-graph.git,Graph-all),
  Project(
    id: Graph-all, 
    base: /Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb, 
    aggregate: List(
      ProjectRef(git://github.com/opyate/scala-graph.git,Graph-core),
      ProjectRef(git://github.com/opyate/scala-graph.git,Graph-json)), 
    dependencies: List(), 
    delegates: List(), 
    configurations: List(compile, runtime, test, provided, optional))), 

  (ProjectRef(git://github.com/opyate/scala-graph.git,Graph-core),
  Project(
    id: Graph-core, 
    base: /Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb/core, 
    aggregate: List(), 
    dependencies: List(), 
    delegates: List(), 
    configurations: List(compile, runtime, test, provided, optional))), 

  (ProjectRef(git://github.com/opyate/scala-graph.git,Graph-json),
  Project(
    id: Graph-json, 
    base: /Users/juanuys/.sbt/staging/9c3cb5d06e75e0970bcb/json, 
    aggregate: List(), 
    dependencies: List(ResolvedClasspathDependency(ProjectRef(git://github.com/opyate/scala-graph.git,Graph-core),None)), 
    delegates: List(), 
    configurations: List(compile, runtime, test, provided, optional))))
```

**Comment by opyate** [comment]
Hi Mikko,

fix #126, now from my branch opyate:issue_126_idiomatic

Same fixes as before:
- Recursively add modules of multi-module dependency
- Add doc for multi-module project

...but with a more idiomatic approach.

Also, the method have been renamed from `getTuplesFromProject` to `getProjectList`, which makes more sense.

Also, next time I'll use [hub](http://stackoverflow.com/questions/4528869/how-do-you-attach-a-new-pull-request-to-an-existing-issue-on-github), because my 2 pull requests have created 2 new issues (which I've closed).
